### PR TITLE
ssh: fix relogin with jumphosts

### PR DESCRIPTION
**Pull request by awly**
Several fixes to make `tsh ssh -J leaf.proxy.com` work if the root cert
is missing/expired.

### CODE REVIEW

**Comment by awly** [comment]
@andrejtokarcik after a bit of tweaking, this change makes the following work:
```
# simulate expired session
$ tsh login --proxy=root.example.com --insecure
...
$ rm ~/.tsh/keys/root.example.com/awly-ssh/*

$ tsh ssh --insecure -J leaf.example.com:3023 leaf-node
```

I'm not confident that all of the changes are safe though, WDYT?

File: lib/client/keystore.go [hunk 1]
```diff
@@ -305,6 +305,7 @@ func (fs *FSLocalKeyStore) GetKey(idx KeyIndex, opts ...CertOption) (*Key, error
 		Priv:     priv,
 		TLSCert:  tlsCert,
 		TrustedCA: []auth.TrustedCerts{{
+			ClusterName:     idx.ClusterName,
```
Comment (andrejtokarcik) [inline lib/client/keystore.go:308]: I don't think this is entirely correct. `TrustedCerts.ClusterName` should correspond to the cluster from which the host CA certs were obtained (which should be the root cluster) but `idx.ClusterName` is often set to a leaf cluster's name.

File: lib/client/api.go [hunk 2]
```diff
@@ -1060,13 +1072,24 @@ func (tc *TeleportClient) LoadKeyForCluster(clusterName string) error {
 func (tc *TeleportClient) LoadKeyForClusterWithReissue(ctx context.Context, clusterName string) error {
 	err := tc.LoadKeyForCluster(clusterName)
 	if err == nil {
+		// TODO(awly): is this always correct?
+		//
+		// Reset client AuthMethods, in case they were populated via relogin
+		// and would force us to use the root cert.
+		if len(tc.AuthMethods) > 0 {
+			tc.AuthMethods = nil
+		}
```
Comment (andrejtokarcik) [inline lib/client/api.go:1081]: I don't think this is needed at all. `tc.AuthMethods` (= `tc.Config.AuthMethods`) is never populated via relogin. It's always set explicitly alongside with `SkipLocalAuth = true`, i.e. `tc.AuthMethods` is never really involved in typical login scenarios. It's used either from `tsh` (when using an identity file) or the web terminal client.

File: lib/client/api.go [hunk 3]
```diff
@@ -462,7 +462,19 @@ func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error)
...
+		errViaJumphost := err
+		// ActivateKey re-fetches the list of CAs from auth server. If
+		// JumpHosts was pointing at the leaf cluster, this could've caused the
+		// above error. Try to ActivateKey without JumpHosts to force it to use
+		// the root cluster.
+		if err := tc.WithoutJumpHosts(func(tc *TeleportClient) error {
+			return tc.ActivateKey(ctx, key)
+		}); err != nil {
+			return trace.NewAggregate(errViaJumphost, err)
+		}
```
Comment (andrejtokarcik) [inline lib/client/api.go:477]: Even though this should work, I don't like this general approach. `WithoutJumpHosts` is wonky, we should focus on when jumphosts *are* supposed to be used, i.e. take a `WithJumpHosts` approach.

Also if we're going down this way, this should probably be put directly into `TeleportClient.UpdateTrustedCA`, not ad hoc in `RetryWithRelogin`.

File: lib/client/api.go (hunk 2 above)
Comment (awly) [reply lib/client/api.go:1069]: It's populated in https://github.com/gravitational/teleport/blob/fbae7ad508d0f30aacbfb2bf5a3878e1c448439b/lib/client/api.go#L472-L481

Here's what happens when relogin triggers in `RetryWithRelogin`:
- user logs in
- root cluster cert is issued
- root cluster cert is set in `tc.Config.AuthMethods`
- on next attempt, `ConnectToProxy` uses `tc.Config.AuthMethods[0]` first
- this cert is accepted by the leaf proxy, because CA validation succeeds
- the leaf proxy ends up routing the connection to `node@root` instead of `node@leaf` because of the cert cluster name extension
- the node connection fails because `leaf` can't forward requests to `root`

Because node connection happens after a successful `ConnectToProxy`, the client doesn't try any other `tc.authMethod`s.

File: lib/client/api.go (hunk 3 above)
Comment (awly) [reply lib/client/api.go]: yeah, it's pretty ugly.
the whole client library is a spaghetti and I don't know where jumphosts really matter and where not.
moved the retry into `ActivateKey` for now.

File: lib/client/keystore.go (hunk 1 above)
Comment (awly) [reply lib/client/keystore.go]: good catch!
added a loop to properly split `tlsCA` into individual `auth.TrusterCert` based on the cluster name from the CN.

File: lib/client/api.go (hunk 2 above)
Comment (andrejtokarcik) [reply lib/client/api.go:1069]: Right I didn't notice this occurrence.

This assignment to `tc.Config.AuthMethods` in `RetryWithRelogin` is just an optimization? Wouldn't it be easier to drop this assignment, and rely on the agent auth methods only...?

File: lib/client/keystore.go (hunk 1 above)
Comment (awly) [reply lib/client/keystore.go]: Removed this change for now, turns out it's not needed.

**Review by andrejtokarcik — APPROVED** [review]

File: lib/client/api.go (hunk 2 above)
Comment (awly) [reply lib/client/api.go]: You're right.
I was afraid to remove it, assuming something relies on the AuthMethods update.
But removing it doesn't seem to break anything on relogin (based on manual testing), so let's do that.

Worst case, we should catch any issues during the testplan.

File: lib/client/api.go [hunk 4]
```diff
@@ -469,15 +469,6 @@ func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error)
...
 	}
-	// Override client's auth methods, current cluster and user name
-	authMethod, err := key.AsAuthMethod()
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	// After successful login we have local agent updated with latest
-	// and greatest auth information, setup client to try only this new
-	// method fetched from key, to isolate the retry
-	tc.Config.AuthMethods = []ssh.AuthMethod{authMethod}
```
Comment (r0mant) [inline lib/client/api.go:480]: Why did you remove this?

**Review by r0mant — APPROVED** [review]
Code lgtm but a question about the removed block.

**Review by russjones — APPROVED** [review]
Bot.

File: lib/client/api.go (hunk 4 above)
Comment (awly) [reply lib/client/api.go:480]: This messes up the relogin flow.

`tc.Config.AuthMethods` is used as the first choice when connecting to a proxy over SSH.
If you were originally connecting to a leaf cluster (as in `tsh ssh -J leaf.proxy.com somenode`) and relogin kicked in, you're be logged in via the root cluster (which is correct).
But after you've re-logged through the root cluster `tc.Config.AuthMethods` would have an SSH cert for the root cluster, not for the leaf.
The cert caching and loading logic that @andrejtokarcik added correctly adds a leaf cert into the `localAgent`, but it gets overriden by `tc.Config.AuthMethods`.
